### ISSUE TITLE
[RFC] DataLoader architecture updates and TarDataset implementation

### ISSUE THREAD

**Issue by VitalyFedyunin** [pytorch/pytorch#49440]
# Problem statement
This proposal aims to construct a modular, user-friendly, and performant toolset to address the ambiguous activity referred to as “dataloading” within PyTorch, a simplification attributable to the indivisibility of the DataLoader abstraction prescribed today. In reality, “dataloading” is a diverse set of operations that should be supported by extensible building blocks, out of which the present abstractions, and far more could be easily built. Some typical needs which are scarcely supported in the present implementation include:

- Lazy loading - Users want to point PyTorch to a remote data source (e.g. HTTP, S3, GCP, Azure, Manifold, Hive) and iterate over the contents without downloading the entire dataset, ideally only downloading samples as-soon-needed. Further, If a user writes such code for a remote storage platform, there is no natural place to contribute it for public or private reuse.

- Structured data, heterogeneous storage

  - There are hundreds of ways to store a single structured dataset, each requiring a custom or highly configured DataLoader. Users want to take advantage of modularity and not reimplement complete DataSets over and over again. Suppose we have a simple dataset of sample:label pairs of images and ints. There are a number of dimensions whose outer product enumerates the possible storage formats for this data, each requiring a distinct (or specifically configured) DataLoader:

    - Primitive formats - Are images stored in an image storage format (e.g. one of these), as a tensor (.pt), as a serialized (e.g. pickle, json) Python data structure, etc.?
    - Grouping - Are pairs grouped together by directory, by an archive format (tar, HDF5), by a serialization format (e.g. json, protobuf), by common file string (e.g. image00012321.jpg, label00012321.txt), by meaningful filenames (e.g. image_00023423_dog.jpg), by contents of file headers, by pickled Python data structures, etc? Are filenames otherwise meaningful? Are file headers otherwise meaningful?
    - Sharding - Is the dataset partitioned for performance reasons into arbitrary groups, each containing grouped pairs? In which grouping format (e.g. directories, tar, arrow, parquet, HDF5)?
    - Compression - Are files or groups compressed or binarized, e.g. gz, zip, protobuf?
    - Locale - Are the files local, remote via custom request format (e.g. proprietary data, public REST API, kaggle dataset), on an http server, in cloud object storage?
  - The above example is only for an extremely simple data structure case. The reality of data is often dramatically more heterogeneous and complex (e.g. variable-length lists of bounding box points and strings in object detection, highly nested structures of user or product features in ranking).
  - Further, users want to find or contribute ops to decode specific file types (e.g. HDF5) and accelerated kernels (e.g. GPU mp3 decoding).
  - Given PyTorch maintains decoders for many storage formats, users want more powerful top-level abstractions, such as simply pointing PyTorch to a local or remote directory and receiving an iterator over best-efforts deserializations of the files within.
- Shuffling - Users want control over when shuffling occurs within “Dataloading.” It often makes a big performance and accuracy difference whether samples within a shard are shuffled, samples are globally shuffled, shards are shuffled, etc.
- Pipelining and parallelism - Users want to be able to pipeline their loading and preprocessing (rather than make multiple CPU passes, for example), specify a number of workers to read and preprocess data, and not worry about whether reading, preprocessing, or model execution are starved. This can include asynchronous processes which prefetch data to feed to others.

Tensorflow addresses many of the above needs with their TFRecord, dramatically simplifying the problem by taking a strong opinion of the data format with which Tensorflow works best. This has been extremely successful from a performance perspective. However, by prescribing a single storage format, all others are demoted, and the diversity of data needs and entrenched formats made ubiquitous adoption of TFRecord for storage practically impossible. We’ve heard directly from users that they do not want to be forced into a single first-class format, and the public datasets (which Google rehosts in TFRecord), tend to agree (by completely disagreeing on format). For this reason, we prefer extensibility over prescription, wherein we provide performant support for a basic set of formats in-tree (e.g. Hive and Manifold internally, tar shard and Arrow externally) but users can plug in modular extensions for new formats easily.

## Underlying DataLoader Issues
Beyond the needs described above, the existing DataLoader is also a frequent source of user requests and github issues. Such feedback includes, but is not limited to:
- Fork and general multi-processing memory usage patterns - There are multiple reports in GitHub that users are confused about how Fork’s copy-on-write and Python’s object counting work together, and that leading to OOMs. And Pytorch users shop for custom solutions as separate list management processes, or sharing binary segments etc.
- Threading vs Multiprocessing - Different use cases require one or the other. For example, threading generally performs better while multiprocessing works better with third-party libraries with non-threadlocal state.
- Overcomplication of solutions - TarDataset requires custom shuffling and sampling implemented as Datasets, while our built-in solution requires altering the DataLoader. It would be best to separate data processing (reordering included) from process management.
- Multiprocessing support - Today, proper pre-fetching is not possible due to the synchronous nature of Datasets. In order to bypass this, users must implement custom multiprocessing-enabled Datasets and DataLoaders themselves.
- Manual sharding - Sharding is increasingly becoming a mandatory feature, allowing better multiprocessing and distributed execution. Currently users must implement it manually.

Finally, the ubiquity of the DataLoader necessitates strong backward compatibility. For this reason we do not plan to deprecate any existing functionality, but in some cases may offer a more modern way of doing things.

# Solution

Break down Dataset into smaller components reducing logic to a queue of data-in and a queue of data-out.  

DataLoader observes the acyclic graph of DataSets and provides the necessary level of parallelism using multiprocessing and multithreading.

Bear in mind that even if we use IteratableDataset in examples below, all this also applicable to MapDataset

### Separating by smaller datasets and connecting them together

```python
class ListFilesIteratableDataset(IteratableDataset):
   #...
   def __iter__(self):
       # yield file_names
 
class LoadFilesIteratableDataset(IteratableDataset):
   def __init__(self, listfiles_ds):
       self._listfiles_ds = listfiles_ds
       # ...
  
   def __iter__(self):
       for file_name in listfiles_ds:
           yield (file_name, load_file(file_name))
```

Will allow us to simplify datasets code and make them reusable across various implementations (for example ImageFolder and TarDataset). Also necessary in case of moving memory consuming datasets into separate processes.

### Turning IteratableDataset and MapDataset into AsyncIteratableDataset and AsyncMapDataset

Multiprocessing/threading support makes us prefer async_next over __next__ function. Key difference is that async_next might throw NotAvailable exception, meaning that data is not yet available and should be requested again with async_next.

Datasets which implements only async_next can be easily used as standard datasets because parent class provides necessary API:

```python
class AsyncIteratableDataset(IteratableDataset):
   def __iter__(self):
       return self
 
   def __next__(self):
       while True:
           try:
               return self.async_next()
           except StopIteration:
               raise StopIteration
           except NotAvailable:
               time.sleep(DELAY)
               EventLoop.iteration()
              
   def async_next(self):
       raise NotImplemented
```

Existing synchronous datasets can be turned into async datasets using helper function: 

```python
def EnsureAsyncNextDataset(validated_dataset):
   if not isinstance(validated_dataset, IteratableDataset):
       raise Exception('Not IteratableDataset')
   if isinstance(validated_dataset, AsyncIteratableDataset):
       return validated_dataset
   if not hasattr(validated_dataset, '_as_iterator'):
       setattr(validated_dataset, '_as_iterator', None)
   if not hasattr(validated_dataset, 'async_next'):
       def async_next(self):
           if self._as_iterator is None:
               self._as_iterator = iter(self)
           return next(self._as_iterator)
       setattr(validated_dataset, 'async_next', async_next)
       validated_dataset.async_next = types.MethodType(async_next, validated_dataset)
   return validated_dataset
```

Combination of two approaches will allow a mix of old-style datasets and new async datasets.

As next_async does not guarantee results to be returned, it can be used to schedule requests ahead:

```python
class PrefetcherAsyncIteratableDataset(AsyncIteratableDataset):
   def __init__(self, source_ds, buffer_size = 10):
       self._souce_ds = source_ds
       self._buffer_size = buffer_size
       self._buffer = []
       self._source_depleted = False
 
   def async_next(self):
       if not self._source_depleted:
           while len(self._buffer) < self._buffer_size:
               try:
                   data = self._souce_ds.async_next()
               except NotAvailable:
                   # break or put more requests, depends from implementation
                   break
               except StopIteration:
                   self._source_depleted = True
                   break
               self._buffer.append(data)
       if len(self._buffer):
           data = self._buffer.pop(0)
           return data
       else:
           if self._source_depleted:
               raise StopIteration
           else:
               raise NotAvailable
```

Similar approach will be applied to MapDataset with async_get(id).

### Connecting blocks with queues

Having all datasets as asynchronous, allows to connect them with a couple of queues.

For example in multiprocessing version, sub process main loop can look like this:

```python
def IteratableDatasetToQueuesLoop(source_dataset, req_queue, res_queue):
   steps = 0
   EventLoop.enabled = False
   for _ in IteratableDatasetBehindQueues(source_dataset, req_queue, res_queue, raise_stop=True):
       steps += 1
       time.sleep(DELAY)
       pass
 
def IteratableDatasetBehindQueues(source_dataset, req_queue, res_queue, raise_stop = False):
   source_dataset = EnsureAsyncNextDataset(source_dataset)
   while True:
       try:
           req_queue.get(block = False)
       except:
           yield True
           continue
       while True:
           try:
               value = source_dataset.async_next()
           except NotAvailable:
               yield True
               continue
           except StopIteration:
               res_queue.put(StopIteration())
               if raise_stop:
                   raise StopIteration
               else:
                   yield True
               continue
           res_queue.put(value)
           yield True # Returns control
           Break
```

When main process can transparently access this dataset with simple wrapper:

```python
class QIteratableDataset(AsyncIteratableDataset):
   def __init__(self, request_queue, response_queue, response_wait_time = 0.00001):
       self._req_q = request_queue
       self._res_q = response_queue
       self._req_sent = False
       self.counter = 0
       self._stop_iteration = False
       self._response_wait_time = response_wait_time
       
   def async_next(self):
       if self._stop_iteration:
           raise Exception('next called after receiving StopIteration')
       if not self._req_sent:
           self._req_q.put(self.counter)
           self.counter += 1
           self._req_sent = True
       try:
           value = self._res_q.get(block = True, timeout = self._response_wait_time)
       except:
           raise NotAvailable
       self._req_sent = False
       if isinstance(value, StopIteration):
           self._stop_iteration = True
           raise StopIteration
       return value
```

Allow to send Dataset into separate process by few lines of code:
 
```python
req_queue = multiprocessing.Queue()
res_queue = multiprocessing.Queue()
p2 = multiprocessing.Process(target=IteratableDatasetToQueuesLoop, args=(source_dataset, req_queue, res_queue))
p2.start()
separated_source_dataset = QIteratableDataset(req_queue, res_queue)
```

Please note, that only one request in the queue, is an implementation restriction and not enforced by design.

### DataLoaderQueue

The above examples using standard multiprocessing Queue, but it is not the best choice (performance-wise) in some cases and not working in others. Instead we suggest to replace it with higher abstraction DataLoaderQueue.

DataLoaderQueue  - used to pass data between elements of a pipeline inside a single thread, between threads, between processes, in distributed env. DataLoader will replace queue with best for the moment implementation, but they all should follow next requirements:

- Non-blocking
- Guaranteed delivery
- Guaranteed no duplicates
- Guaranteed order
- Customizable length
- Queue is always between TWO processes/threads

API: 
- `def get(blocking=True)` - returns any python structure, or raises NotAvailableException, or raises QueueError
- `def put(data, blocking=True)` - data is any Python structure, may raise QueueError

DataLoaderQueue implementation also defines ‘serialization’ technique, from simple pass object reference inside the same thread to IPC calls and full object serialization to be passed via network. 

### Users API

Datasets should work as standard iterators (or implement __get__item__) outside of DataLoader.

```python
numbers_ds = ds.NumbersIteratableDataset() # Returns range of integers
ds1, ds2, ds3 = MultiplyIteratableDatasetList(numbers_ds, 3) # Creates 3 copies of input data
def mult100(x):
   return x * 100
ds2_modified = CallableIteratableDataset(ds2, mult100)
def mult111(x):
   return x * 111
ds3_modified = CallableIteratableDataset(ds2, mult111)
joined_ds = GreedyJoinIteratableDataset(ds1, ds2_modified, ds3_modified)
 
for i in iter(joined_ds):
   print(i) # 0 0 0 1 100 111 222 200 2 ......
```

DataLoader output should be exactly the same, but different pieces of graph might be executed as separate threads/processes.

```python
for i in DataLoader(joined_ds):
   print(i) # 0 0 0 1 100 111 222 200 2 ......
```

### Sharding

Sharding should be implemented on the framework level and hidden from DataSet users. DataSet developers will get control over sharding settings and running configurations. DataLoader will decide how to split DataSet into shards and run configuration.

DataSet blocks will provide information to the DataLoader if they support sharding via dataset.is_shardable(). If a function is not defined Dataset will be considered as non-shardable. 
DataLoader will callback DataSet objects with sharding settings using dataset.sharding_settings(total_num_of_shards, id_of_shard).

Example:

```python
list_files_ds = ListFilesMapDataset(root = '.') * marked as non shardable
load_bins_ds = LoadFilesMapDataset(list_files_ds) * marked as shardable
decode_images_ds = DecodeImagesMapDataset(load_bins_ds) * marked as shardable
transform_ds = TransformImagesDataset(decode_images_ds) * marked as shardable shuffle_ds = ShuffleMapDataset(transform_ds) * marked as non shardable
sampler_ds = SamplerIteratableDataset(shuffle_ds) * marked as non shardable
```

### Individual Process (Thread)

Situations like prefetching and large non-forkable arrays require to spawn separate processes for a Dataset. DataSet blocks will provide information to the DataLoader if they are recommended to be executed as separate processing via dataset.is_separate_process().

### Lazy Initialization

In some cases it is inefficient to initialize DataSet data before usage. For example, we need to postpone loading a full list of files before forking out a file scanner. For this purpose lazy_init function will be called prior to any __len__, __get_item__, __iter__ operators. 

### Moving Sampler and Shuffle from DataLoader into separate Datasets 

DataLoader should not care about any data logic (including sampling and shuffle). We are planning to create Samplers and Shuffle datasets for each existing logic as well as a wrapper around existing Sampler classes: ds =SamplerDataset(input_ds, old_sampler_class_obj)

### Moving Collate functions from DataLoader into separate Datasets

We are going to move collate logic out of DataLoader and implement it as IterableDataset, it will accept old collate functions as argument or can be rewritten entirely.

Example: https://github.com/pytorch/pytorch/pull/48933

```python
batch_ds = ds.BatchNumbersIteratableDataset() # Returns batch of integers [1,2,3],[4,5,6],..
default_collated_ds = CollateDataset(batch_ds) # use original default collate function
for i in DataLoader(default_collated_ds):
   print(i) # tensor([1, 2, 3]), tensor([4, 5, 6]), ...
 
def collate_fn(batch):
    sum = batch[0] + batch[1] + batch[2]
    return torch.tensor(sum, dtype=torch.float)
collated_ds = CollateDataset(batch_ds, collate_fn=collate_fn)
for i in DataLoader(default_collated_ds):
   print(i) # tensor([6.]), tensor([15.]), ...
```

### Reproducibility and randomness

Should be part of DataLoader implementation, to be able to define random seed in case of various parallelization techniques.

Async operations also introduce non-determinism of order, so we would need to implement a DataLoader attribute to order of async calls fulfillments and to guarantee order determinism.

### To Do

This document doesn’t touch the problem of varying batch size for different phases of processing. It is archivable by passing a list of objects into the queue and will be considered at the phase of queue implementation. However it is better to put code example here.

This document doesn't cover distributed training in detail. We are going to extend on this topic using additional sharding parameters and queue implementations.

### Considerations

User defined sharding was considered unnecessary at the early stages, however, nothing in the proposed architecture prevents from implementing it later.

CPP implementation was considered as non-flexible. However, nothing prevents users from creating Datasets with CPP internals.

Torchscript can be used inside of Datasets, but we are not limited to it.

Arrow/Proto/… can be used to pass data between Datasets. 

Error Tracing?

C++

**Comment by pritamdamania87** [comment]
@VitalyFedyunin Is it possible to update the RFC with a simple example of what the dataloader would look like while training with DDP where we need to shard data?

**Comment by ezyang** [comment]
Reminder that https://github.com/pytorch/rfcs/ is a thing, and might be easier to do in depth comments on the RFC

**Comment by VitalyFedyunin** [comment]
Answering sharding question. Let say we have dataset with access to some database of N elements.
We need to shard it and yield only 1/X of elements from this database, non overlapping with other 
shards.

```python
class IterateableDataset:
    pass

class NumbersIteratableDataset(IterateableDataset):
    def is_shardable(self):
        return True

    def __init__(self, range = 100):
        self.range = range

    def sharding_settings(self, total_num_of_shards, id_of_shard, seed = 0):
        # Called before __iter__
        self.total_num_of_shards = total_num_of_shards
        self.id_of_shard = id_of_shard
        self.seed = seed

    def __iter__(self):
        for i in range(self.id_of_shard, self.range, self.total_num_of_shards):
            yield (i + self.seed) % self.range
```

Now lets assume X == 3, so we have three consumers (and it doesn't matter how they are separated by physical/logical nodes).
For example it can be 1 on shard inside one process and two shards on the different machine.

```python
# First
number_of_shards = 3
shard = NumbersIteratableDataset()
shard.sharding_settings(number_of_shards, 0)
```
```python
# Second
number_of_shards = 3
shard_0 = NumbersIteratableDataset()
shard_0.sharding_settings(number_of_shards, 1)
shard_1 = NumbersIteratableDataset()
shard_1.sharding_settings(number_of_shards, 2)
```
So we basically need to share between machines only total number of shards and some cross system incremental id 

```python
for i in iter(shards_1):
    print(i)
# 2, 5, ...., 98
```

It is likely that we will need to have access to other data on next epoch, for this we can introduce seed or epock number to the sharding_settings

```python
# Next epoch 
for i, shard in enumerate(shards):
    shard.sharding_settings(number_of_shards, i, seed = 777 * epoch_number )

for i in iter(shards_2):
    print(i)
# 79, 82 .... 97, 0, 3, ... 75
```

**Comment by VitalyFedyunin** [comment]
The alternative option available when we have only legacy dataset without sharding support, nesting it with something like sharding filter will help

```python
class OldDataset:
    def __iter__(self):
        for x in range(100):
            yield "str%d" % x

class ShardedFilterIterableDataset:
    def __init__(self, source_ds):
        self.source_ds = source_ds

    def sharding_settings(self, total_num_of_shards, id_of_shard, seed = 0):
        # Called before __iter__
        self.total_num_of_shards = total_num_of_shards
        self.id_of_shard = id_of_shard
        self.seed = seed

    def __iter__(self):
        for i, value in enumerate(self.source_ds):
            if i % self.total_num_of_shards == self.id_of_shard:
                yield value

number_of_shards = 3
legacy_ds = OldDataset()
sharded = ShardedFilterIterableDataset(legacy_ds)
sharded.sharding_settings(number_of_shards, 2)

for x in sharded:
    print(x)
# str2, str5 ....
```

**Comment by bgedik** [comment]
Two features I’d suggest:
 * Checkpointing 
 * Uneven end of data handling (for distributed training)

**Comment by VitalyFedyunin** [comment]
Can you please elaborate on the "Uneven end of data handling (for distributed training)" topic. What exactly you are looking for?

Good point about checkpointing, just to be clear, we want to be able to stop straining at the 'middle' of epoch (data loop) and resume it after a cold restart.

**Comment by bgedik** [comment]
say you have n data loader instances, in many distributed training scenarios if one instance runs out of data then all has to run out of data. Otherwise the training hangs.

**Comment by carmocca** [comment]
>     * Uneven end of data handling (for distributed training)

Similarly, supporting https://github.com/pytorch/pytorch/issues/25162#issuecomment-731361982

**Comment by jspisak** [comment]
cc @jph00 for viz and any inputs. Also anyone else from the fast.ai community you think should comment here?

**Comment by nairbv** [comment]
nit: IteratableDataset -> IterableDataset

**Comment by nlgranger** [comment]
Here are a few thoughts to contribute to this discussion:
- Pre-processing cost varies greatly across domains and buying one-size-fits-all GPU nodes is complicated. Consequently, distributed data loading ought to be a targeted use-case for the data loader re-design so that this variability can be off-loaded to CPU-only servers.
- I might be missing a point, but why is sharding a topic by itself? Isn't it easy to manage it in a similar fashion to DDP sampler? It's difficult to anticipate all possible use-cases so unless it requires a lot of boilerplate code from users it's probably best to leave this outside the core of pytorch dataloader. Maybe just provide a few utilities and wrappers?
- Reading the above posts didn't give me a clear idea on how the ordering of requested items is preserved.
- What motivates the need for async item access? Workers idling during IO? Is this a real-world problem that can't be solved by adding more workers?

I have made a few experiments with data loading before, here is my modest feedback and sorry for the self-advertisement:
- Iterable and indexable datasets are more easily handled via separate code-bases and APIs, there might be a bit of code duplication but it will simplify both implementations greatly. The access to items and supported data loading strategies vary too much to uniformize them. For instance shuffling an iterable dataset requires to build a buffer where items are shuffled.
- Sampling and multi-processing don't play well together. A sampler API is needed to standardize when the upcoming item to fetch are requested (ex: capture item i + p when i is completed or iterate asynchronously on the sampler), how to make the sampler multiprocess-friendly?
- "Moving Sampler and Shuffle from DataLoader into separate Datasets": [see example here](https://seqtools-doc.readthedocs.io/en/latest/examples/dataloader.html) (jump to make_sequence method).

Extra minor remarks:
- How about bringing people who know about the rdma stuff in the discussion about distributed data-loader?
- About the OOM due to COW, persistent workers which start when the dataloader should help, it's also more intuitive.
- About OOM due to shared memory, maybe use a fixed shm memory pool. [Check the same example](https://seqtools-doc.readthedocs.io/en/latest/examples/dataloader.html) to see that in action, it is actually implemented by [this function](https://seqtools-doc.readthedocs.io/en/latest/reference.html#seqtools.prefetch) which uses refcounting hooks to track free memory.
-  I don't think a use-case where the order of items is not important should be considered because it doesn't encourage reproducibility
- Python 3.8 introduced features to help implement Apache Arrow features with pickle: [PEP 574](https://www.python.org/dev/peps/pep-0574/), [sample use-case](https://github.com/nlgranger/SeqTools/blob/7ee3aa3abb7bb82e3377ba8180c3abccb859e17e/seqtools/evaluation.py#L216) which is also used by the dataloader example.

**Comment by npuichigo** [comment]
> Answering sharding question. Let say we have dataset with access to some database of N elements.
> We need to shard it and yield only 1/X of elements from this database, non overlapping with other
> shards.
> 
> ```python
> class IterateableDataset:
>     pass
> 
> class NumbersIteratableDataset(IterateableDataset):
>     def is_shardable(self):
>         return True
> 
>     def __init__(self, range = 100):
>         self.range = range
> 
>     def sharding_settings(self, total_num_of_shards, id_of_shard, seed = 0):
>         # Called before __iter__
>         self.total_num_of_shards = total_num_of_shards
>         self.id_of_shard = id_of_shard
>         self.seed = seed
> 
>     def __iter__(self):
>         for i in range(self.id_of_shard, self.range, self.total_num_of_shards):
>             yield (i + self.seed) % self.range
> ```
> 
> Now lets assume X == 3, so we have three consumers (and it doesn't matter how they are separated by physical/logical nodes).
> For example it can be 1 on shard inside one process and two shards on the different machine.
> 
> ```python
> # First
> number_of_shards = 3
> shard = NumbersIteratableDataset()
> shard.sharding_settings(number_of_shards, 0)
> ```
> 
> ```python
> # Second
> number_of_shards = 3
> shard_0 = NumbersIteratableDataset()
> shard_0.sharding_settings(number_of_shards, 1)
> shard_1 = NumbersIteratableDataset()
> shard_1.sharding_settings(number_of_shards, 2)
> ```
> 
> So we basically need to share between machines only total number of shards and some cross system incremental id
> 
> ```python
> for i in iter(shards_1):
>     print(i)
> # 2, 5, ...., 98
> ```
> 
> It is likely that we will need to have access to other data on next epoch, for this we can introduce seed or epock number to the sharding_settings
> 
> ```python
> # Next epoch 
> for i, shard in enumerate(shards):
>     shard.sharding_settings(number_of_shards, i, seed = 777 * epoch_number )
> 
> for i in iter(shards_2):
>     print(i)
> # 79, 82 .... 97, 0, 3, ... 75
> ```

What if the shards can not be evenly dispatched to different workers (maybe GPUs) or the number of items in shards are different? I saw in pytorch 1.7, join op can be used to address the problem of uneven dataset. But in the case of sharing, the last worker or GPU may still have a whole shard to be consumed when the others are idle. Does this need more consideration?

**Comment by npuichigo** [comment]
any update? @VitalyFedyunin

**Comment by ppwwyyxx** [comment]
Reposting this sharding question from the google doc version:

> `dataset.sharding_settings` changes the internal state of `dataset`. What do you think about the alternative which lets `dataset.sharding(...)` returns a new `dataset` instead?
I'm thinking about a situation where a dataset is sharded into multiple levels, e.g. in a setup with A GPUs, each with B dataloader worker processes, each worker uses C threads. The alternative seems easier to me for such cases, but with the current proposal I'm not sure how to create multiple levels of sharding.

**Comment by npuichigo** [comment]
> Reposting this sharding question from the google doc version of this RFC:
> 
> > `dataset.sharding_settings` changes the internal state of `dataset`. What do you think about the alternative which lets `dataset.sharding(...)` returns a new `dataset` instead?
> > I'm thinking about a situation where a dataset is sharded into multiple levels, e.g. in a setup with A GPUs, each with B dataloader worker processes, each worker uses C threads. The alternative seems easier to me for such cases, but with the current proposal I'm not sure how to create multiple levels of sharding.

Can u share the link? Is that public?

**Comment by VitalyFedyunin** [comment]
Hello!

Doc URL is https://docs.google.com/document/d/1ILJQl1QFUWFRbmFW50askG5l4t_8xcvSMbIwHFK6DoU/edit?usp=sharing
but please refer to this PR as the most up-to-date version, as keeping all in sync is a nightmare. And sorry for the delay, I aggregating multiple feedbacks now into a singular update which is the next Tuesday.

**Comment by tmbdev** [comment]
It looks like an ambitious redesign; I'm looking forward to its implementation. In particular, cleaning up the DataLoader and multiprocessing facilities will be very useful.

I want to point people at related resources that we have developed for these kinds of representations and processing pipelines.

- The webdataset library (github.com/tmbdev/webdataset) is a library of standard IterableDataset implementations that interoperate with the existing PyTorch library and provide functionality similar to this design.
- WebDataset is also a file format, namely storing datasets in POSIX tar archives with a simple naming convention. If you store your datasets in WebDataset format, you can read them efficiently with the current webdataset library, and you will later be able to read them natively with the new TarIterator
- The [tarp](github.com/tmbdev/tarp) program is a command line program for quick and easy processing of large datasets in WebDataset format; it's a kind of "xargs for tar archives"
- The [AIStore](github.com/nvidia/aistore) server is an open source, high performance storage server that provides infinitely linearly scable I/O performance for WebDataset; it also provides server-side shuffling offline ETL, and online data transformations for WebDataset-style datasets

**Comment by tmbdev** [comment]
A comment about the design and objectives of the project in general. I think from a software engineering point of view, it is nice to have support for all the Python I/O and concurrency facilities: async, threading, queues, and multiprocessing, and to allow for arbitrary sources and sinks. If you have the software develoipment resources, by all means, it's good to implement that.

Keep in mind, though, that any I/O pipeline you write as part of the deep learning job is limited to the CPU cores and PCI bandwidth on the machine with the GPU cards. On 8 or 16 GPU nodes, that greatly limits what you can do.

What we do for large scale training jobs is to run the I/O pipelines on separate nodes from the GPU pipelines. There are two common and easy ways of doing that:

- The [AIStore](http://github.com/nvidia/aistore) server allows on-the-fly processing pipelines to run on the storage server itself; that is, a sharded dataset can be stored on the distributed AIStore server and when you open each shard, the data augmentation and other kinds of data transformations are run transparently by AIStore.
- The [Tensorcom](http://github.com/nvlabs/tensorcom) library lets you write preprocessing, shuffling, batching, and data augmentation pipelines in Python and then sends training batches directly to the GPU nodes using either ZMQ or Direct-to-GPU. The latter gives you 50-100 Gbps network bandwidth _per GPU_ on high-end machines.

Note that this stuff works today and scales linearly (limited only by available hardware).

So, while I'm glad that async and threading are making it into the redesigned I/O pipelines, high performance I/O solutions for large scale deep learning probably just end up running a large number of distributed jobs, each of which is fairly simple and sequential.

**Comment by maxluk** [comment]
Thank you for the write up! It moves forward the design of DataLoader and addresses some common problems, so it's a great effort. It's hard to make conclusions on the solution though without end-to-end examples. The proposal is heavy on architectural principles which is good, but lacks description of how end user will interact with the system and how the end solution will look like. Can you include examples of how user code will look like for at least some of the problems you state in the beginning?
Other comments:
* How APIs for composing data pipelines will look like? What are the goals for the pipelines? Which use cases will they cover? Which transformations are going to be supported out of the box? For which data types?
* Is there any provision to offloading dataloading to external set of CPU nodes? Or is it a non goal?
* How would global shuffling look like at architectural level?
* More details on Caching in addition to mention that there is a functional DataPipe for that?
* More details on modularity? Here, example of how Azure Blob or some other data reader component would integrate into user code would be especially interesting. Or use structured data source for example.

**Comment by VitalyFedyunin** [comment]
Thanks, @tmbdev we are going to put some time and review the potential of adding AIStore and/or TensorCom as DataPipes.

@maxluk we are currently working on the next iteration of the RFC with extra attention to usage examples, your feedback is important and we will take it into account as well. 

Answering other questions:
- yes, we are considering supporting Distributed Data Loading in the future. But can't promise out of box soluting sooner than early 2022.
- Shuffle would be implemented as DataPipe, by default we will provide buffer shuffle.
- Likely we will have simple in-memory caching DataPipe as an example, and help users to develop a set of various caching data pipes.
- I will add more examples on modularity, as we had it in mind during the design phase.

**Comment by tmbdev** [comment]
@maxluk [Here](https://www.youtube.com/playlist?list=PL0dsKxFNMcX59napupGk3cNVzbpx3yU0_) are YouTube videos explaining how shuffling, distribution, caching, and parallelization work for sequential loaders and pipes.

**Comment by npuichigo** [comment]
I saw the recent updates in torch.utils.data.datapipes and tried to utilize that for my own project.

The distributed sharding solution is to add a `SamplerIterDataPipe` after `ListDirFilesIterDataPipe`, and the sampler used in `SamplerIterDataPipe` is used to split the shards to different workers on different gpus.
```python
class DistributedPartitionSampler(Sampler[int]):

    def __init__(self, data_source, *, shuffle=True, seed=0):
        super().__init__(data_source)
        self.data_source = data_source
        self.shuffle = shuffle
        self.epoch = 0
        self.seed = seed

    def __iter__(self):
        if self.shuffle:
            # deterministically shuffle based on epoch and seed
            g = torch.Generator()
            g.manual_seed(self.seed + self.epoch)
            indices = torch.randperm(len(self.data_source), generator=g).tolist()
        else:
            indices = list(range(len(self.data_source)))

        worker_info = get_worker_info()
        if worker_info is None:
            num_workers = 1
            worker_rank = 0
        else:
            num_workers = worker_info.num_workers
            worker_rank = worker_info.id

        if not dist.is_initialized():
            gpu_rank = 0
            gpu_size = 1
        else:
            gpu_rank = dist.get_rank()
            gpu_size = dist.get_world_size()

        skip = max(1, gpu_size * num_workers)
        offset = gpu_rank * num_workers + worker_rank

        shared_index = indices[offset::skip]

        for index in shared_index:
            yield self.data_source[index]

    def set_epoch(self, epoch: int) -> None:
        self.epoch = epoch
```
Here the data source need to be `Sized`, but `ListDirFilesIterDataPipe` using yield and the total number is unknown, so I need to retrieve the file list before partition.
```
pipe = ListDirFilesIterDataPipe('.', masks='*.tar')
pipe = SamplerIterDataPipe(list(pipe), #or [x for x in pipe] because __len__ may throw exception 
                                              sampler=DistributedPartitionSampler)
```
Now I could get different shards on different workers (carefullly set epoch when training)
```
>>> # spawn with 2 GPUs and 3 workers per GPU...(12 shards)
    >>> # GPU-0 WORKER-0 gets [0,6]
    >>> # GPU-0 WORKER-1 gets [1,7]
    >>> # GPU-0 WORKER-2 gets [2,8]
    >>> # GPU-1 WORKER-0 gets [3,9]
    >>> # GPU-1 WORKER-1 gets [4,10]
    >>> # GPU-1 WORKER-2 gets [5,11]
```

**Comment by orena1** [comment]
Probably not exactly in the scope of this PR, but one thing to keep in mind with this refactor is that it will have the ability to add MPI support for the dataloader, so that a GPU could use a cluster of CPUs for data loading.
In my case this was the bottleneck and having this ability would allow me to skip (or do online) many preprocessing steps.

**Comment by vadimkantorov** [comment]
For ease of reimplementation of TF/Google papers, I propose to at least discuss an option of implementing support for TFRecord in core (as one of performant formats)

**Comment by vadimkantorov** [comment]
May be related, stumbled upon a similar I/O library: https://github.com/podgorskiy/DareBlopy

**Comment by tmbdev** [comment]
DareBlopy addresses the small file problem using zip archives, but zip files don't allow streaming and can't guarantee sequential read access. Using tar archives as the underlying storage formats means that datasets can be streamed, that training starts as soon as the first sample has been transferred, that all accesses can be done with large reads, and that no seeks are needed.

In general, if you want high performance, reusable I/O pipelines, it's best to write them using stream processing primitives and assuming only sequential/streaming access to the underlying data ("as if the data is in POSIX tar archive"); such pipelines will then take maximum advantage of any underlying storage format, even if that format is something like zip files or relational databases.

**Comment by npuichigo** [comment]
> it's best to write them using stream processing primitives and assuming only sequential/streaming access to the underlying data ("as if the data is in POSIX tar archive"); such pipelines will then take maximum advantage of any underlying storage format, even if that format is something like zip files or relational databases.

Hi @tmbdev, you said zip files don't allow streaming, but you also said we can assume only sequential access to the underlying data even it's something like zip files. I don't understand that.

I found the implementation of zip reading (https://github.com/pytorch/pytorch/blob/master/torch/utils/data/datapipes/iter/readfilesfromzip.py), and the question is whether this method addresses the issue that zip files don't allow streaming and it's still better to use tar.
```
for data in self.datapipe:
            validate_pathname_binary_tuple(data)
            pathname, data_stream = data
            try:
                # typing.cast is used here to silence mypy's type checker
                zips = zipfile.ZipFile(cast(IO[bytes], data_stream))
                for zipinfo in zips.infolist():
                    # major version should always be 3 here.
                    if sys.version_info[1] >= 6:
                        if zipinfo.is_dir():
                            continue
                    elif zipinfo.filename.endswith('/'):
                        continue

                    extracted_fobj = zips.open(zipinfo)
                    inner_pathname = os.path.normpath(os.path.join(pathname, zipinfo.filename))
                    # Add a reference of the source zipfile into extracted_fobj, so the source
                    # zipfile handle won't be released until all the extracted file objs are destroyed.
                    # Add `# type: ignore` to silence mypy's type checker
                    extracted_fobj.source_zipfile_ref = zips  # type: ignore
                    # typing.cast is used here to silence mypy's type checker
                    yield (inner_pathname, cast(BufferedIOBase, extracted_fobj))
            except Exception as e:
                warnings.warn(
                    "Unable to extract files from corrupted zipfile stream {} due to: {}, abort!".format(pathname, e))
                raise e
```

**Comment by vadimkantorov** [comment]
Also, maybe theoretically ZIP supports some sort of streaming: https://github.com/richgel999/miniz/issues/180
not sure if it's widespread or not

**Comment by tmbdev** [comment]
I'm saying that it's best to write I/O pipelines that only make use of sequential access because such I/O pipelines will give the best performance and will be the most flexible.

If you write your pipeline as:

> list-files | shuffle | ReadFilesFromZip | group | decode | augment | batch

then the `ReadFilesFromZip` will make random accesses to the zip file (in fact, it will do so even if you leave out the shuffle); this is slow in general and requires the entire .zip file to be present on disk. 

When data is stored in a single .tar file, WebDataset uses a different pipeline:

> echo url | open-binary-stream | tar-decoder | group | shuffle | decode | augment | batch

This pipeline can open streams anywhere, doesn't require any local disk storage, and has almost no latency (samples appear almost immediately). Furthermore, you can replace the "open-binary-stream | tar-decoder" with any other sequential data source without affecting the rest of the pipeline.  Of course, if you have a random access data source (like a zip file or a file system), you can also treat that as a sequential data source.

When data is sharded, WebDataset uses this pipeline:

> list-shards bucket-name | shuffle | open-binary-stream | tar-decoder | group | shuffle | decode | augment | batch

This combination of shard shuffling and sample shuffling is the only way I know of to get really high performance I/O for very large datasets.

Zip files have their file directory structure at the very end of the file, so all bytes in a zip file have to be downloaded in order to be able to read it. Your code will first seek to the end of the zip file, read the directory, then seek to each file within the zip file, generating a lot of random accesses.

For most users, tar archives are probably the best technical solution: they are efficient by default, and they prevent users from accidentally using representations that require a lot of local storage or random disk access, both of which cause problems.

**Comment by tmbdev** [comment]
@vadimkantorov In standard zip files, the directory structure containing the definitive file info is at the end of the zip file (in order to make zip files extensible). If a file is deleted from the directory, it is considered removed from the zip file even if the file header and contents still exist within the zip file, so you need the directory in order to know what the actual content of the zip file is. If you ignore the directory structure in a normal zip file, you may get bad data, or duplicate records, or other abnormalities, and worse yet, these may lead to hard to diagnose problems.

You could theoretically limit a deep learning I/O library to just the subset of zip file that are streamable. But in that case, most zip files that users have, use, or generate would not be supported (since they are not guaranteed to have that property), and worse yet, what the I/O library sees might differ from what zip utilities show the user. In other words, "streamable zip files" is a very unusual format and not widely supported.

The advantage of POSIX tar archives is that they are guaranteed to be streaming (they were designed for a sequential medium); every POSIX tar archive will be read exactly according to spec by WebDataset and other I/O libraries. All libraries and tools have exactly the same view of the data. Furthermore, there are many libraries and tools that provide random access to POSIX tar archives when you need it (e.g. for browing data).

We chose tar archives as our standard format after exploring a lot of alternatives, including zip, sqlite3, LMDB, etc. and working in a lot of different environments (local, VMs, MPI, containers, K8s, etc.). Overall, I see a number of disadvantages to using zip files compared to tar files, and no obvious advantages.

**Comment by sbrodehl** [comment]
About a dataset, where sharding results in multiple workers having an uneven amount of data, which can result in "incomplete last batches". Currently, data from different shards is not merged, will this change?

Related issue: #44108

**Comment by npuichigo** [comment]
> Answering sharding question. Let say we have dataset with access to some database of N elements.
> We need to shard it and yield only 1/X of elements from this database, non overlapping with other
> shards.
> 
> ```python
> class IterateableDataset:
>     pass
> 
> class NumbersIteratableDataset(IterateableDataset):
>     def is_shardable(self):
>         return True
> 
>     def __init__(self, range = 100):
>         self.range = range
> 
>     def sharding_settings(self, total_num_of_shards, id_of_shard, seed = 0):
>         # Called before __iter__
>         self.total_num_of_shards = total_num_of_shards
>         self.id_of_shard = id_of_shard
>         self.seed = seed
> 
>     def __iter__(self):
>         for i in range(self.id_of_shard, self.range, self.total_num_of_shards):
>             yield (i + self.seed) % self.range
> ```
> 
> Now lets assume X == 3, so we have three consumers (and it doesn't matter how they are separated by physical/logical nodes).
> For example it can be 1 on shard inside one process and two shards on the different machine.
> 
> ```python
> # First
> number_of_shards = 3
> shard = NumbersIteratableDataset()
> shard.sharding_settings(number_of_shards, 0)
> ```
> 
> ```python
> # Second
> number_of_shards = 3
> shard_0 = NumbersIteratableDataset()
> shard_0.sharding_settings(number_of_shards, 1)
> shard_1 = NumbersIteratableDataset()
> shard_1.sharding_settings(number_of_shards, 2)
> ```
> 
> So we basically need to share between machines only total number of shards and some cross system incremental id
> 
> ```python
> for i in iter(shards_1):
>     print(i)
> # 2, 5, ...., 98
> ```
> 
> It is likely that we will need to have access to other data on next epoch, for this we can introduce seed or epock number to the sharding_settings
> 
> ```python
> # Next epoch 
> for i, shard in enumerate(shards):
>     shard.sharding_settings(number_of_shards, i, seed = 777 * epoch_number )
> 
> for i in iter(shards_2):
>     print(i)
> # 79, 82 .... 97, 0, 3, ... 75
> ```

@VitalyFedyunin Another question about sharding. Since we may use `ListDirFilesIterDataPipe` to yield shards, sometimes we don't know the actual number of shards. So how to deal with this situation.

**Comment by zhuzilin** [comment]
Is there any update on the `MapDataPipe`? I've found that the `IterDataPipe` was already merged into the master branch.

The problem with the `tf.data`-style iterative dataset is that it's hard to implement efficient shard or skip and can only do shuffling in a local pool instead of global shuffling. To be specific, we have to shard the tfrecord filenames instead of sharding the data entries to get a good performance, otherwise `tf.data` has to at least read the header of all records. (Before tensorflow/tensorflow#46358, each shard would even have to read all the data...)

Therefore, I really hope we could have a separate pipeline logic for the map datasets instead of turning the map ones into iteratives.

**Comment by npuichigo** [comment]
@zhuzilin The shuffing issue can be addressed by hierarchical sampling as mentioned in [this paper](http://martin.zinkevich.org/publications/nips2010.pdf) and [chunk dataset](https://github.com/pytorch/pytorch/pull/26547). 
The idea is to ramdonly split shards to GPUs plus using local shuffling. Actually it's what webdataset does.

**Comment by ejguan** [comment]
@npuichigo 

> @VitalyFedyunin Another question about sharding. Since we may use ListDirFilesIterDataPipe to yield shards, sometimes we don't know the actual number of shards. So how to deal with this situation.

I am actually quite curious about why `ListDirFilesIterDataPipe` prevents you having actual number of shards?

**Comment by zhuzilin** [comment]
@ejguan Thank you for your reply:)

I understand there are great advantages to use iter-style datasets. And it will be much easier to deal with all kinds of data formats or data sources if we switch to it. `tf.data` and the underlying io abstraction has proved its power in practice. I just hope that we could have a map-style counterpart to achieve the things that is hard for iter-style.

There are two use cases on my mind at the moment: sharding and caching.

- First, about sharding.

I've been working on improving the performance of `shard` in the tensorflow community recently. As I mentioned in the comment above, because `tf.data` is completely iter-style (from the dataset ops, to the underlying io streams or file abstractions, to tfrecord the format itself), it's hacky to do operations that need to skip records. Though I added a `Skip` api to the iterator base class (tensorflow/tensorflow#40963), we still have to disable the `BufferedInputStream` underneath to avoid reading data that are skipped (https://github.com/tensorflow/tensorflow/pull/46358#issuecomment-790475010). Those require the user to have much knowledges of the framework architecture, otherwise, a `shard(128, i)` will make the io speed 100 times slower.

The workaround by tensorflow community is to shard the file names and read each file without the shard:

```python
d = tf.data.Dataset.from_tensor_slices(filenames)
d = d.shard(num_workers, worker_index)
d = d.flat_map(lambda x: tf.data.TFRecordDataset(x))
```

instead of the intuitive:

```python
d = tf.data.TFRecordDataset(filenames)
d = d.shard(num_workers, worker_index)
```

This work around would fail if the user have uneven-sized tfrecord files or maybe only one large file, which are common in practice.

On the other hand, the map-style dataset can easily achieve sharding with `index = index + num_workers`. And for tfrecord file, we could extract the offset of each record in the dataset ahead and use this metadata to do random read.

- Second, about caching.

In our scenario, the user would save some large dataset on remote file systems like ceph and the speed fluctuation of the FS may affect the training speed a lot. Therefore, we need to cache part of the dataset to local disk to make IO more stable. However, it's hard to implement caching if we only need to store part of the data (The whole dataset is too large (TBs) to be hold in the local disk (GBs)), because we cannot check if a data entry is cached or not. Even if we can, we cannot get the entries that are not cached, since we cannot have random access to the dataset. Therefore, we only implement the caching functionality for pytorch, but not tf at the moment.

Again, I need to emphasis that I agree iter-style is a really nice model for most use cases. I just hope we could have some support for the map-style as well. The current dataset and dataloader model supports map-style and I really loved that. Hope we could find a way to keep on support at least part of that.

There are some comment on your rely.

>Like shuffling, you can still have the global shuffling by using a buffer that can hold the whole dataset, even though it's not recommended.

> Like shuffling, you need to load index or filenames (which will be similar to Iter-style) into memory and then shuffle, the memory overhead is still same as Iter-style.

For the iter-style, one need to get a pool that is the same size of the dataset to do global shuffling, while the map style would only need a function to produce the permutation of the index or directory. An image is several-hundred Ks, while a directory is only a couple of bytes. I don't think they have the same memory overhead.

> Especially for the large-scale dataset, you can't really load everything into memory, even operations over indexes may cause OOM.

I agree. But I think most datasets do not belong to this category... I think it's reasonable to assume most users are facing the datasets at the scale of GBs to TBs instead of the gigantic ones that would take GBs only for index or directory. Even in this case, the map-style can still fallback to have a pool of index, which could hold a lot more records than the data pool of iter-style.

Thank you for your time on this looong comment~

**Comment by npuichigo** [comment]
> @npuichigo
> 
> > @VitalyFedyunin Another question about sharding. Since we may use ListDirFilesIterDataPipe to yield shards, sometimes we don't know the actual number of shards. So how to deal with this situation.
> 
> I am actually quite curious about why `ListDirFilesIterDataPipe` prevents you having actual number of shards?

@ejguan ListDirFilesIterDataPipe uses generator to yield shard file names, and `__len__` is not implemented only if we pass in the length. 
```python
    def __iter__(self) -> Iterator[str] :
        yield from get_file_pathnames_from_root(self.root, self.masks, self.recursive, self.abspath)

    def __len__(self):
        if self.length == -1:
            raise NotImplementedError
        return self.length
```

So my question is how sharding works if all the `IterDataPipe` used are of unknown length. Or does it mean we may need pre-compute the number of (shard) files and then inject that to shard settings?

**Comment by nairbv** [comment]
Re: Uneven end of data handling

A common solution is an extra layer of indirection between "shards" and workers (which I guess would be instances of `NumbersIteratableDataset()` in the example).  E.g., instead of:
```
# First
number_of_shards = 3
shard = NumbersIteratableDataset()
shard.sharding_settings(number_of_shards, 0)
# Second
number_of_shards = 3
shard_0 = NumbersIteratableDataset()
shard_0.sharding_settings(number_of_shards, 1)
shard_1 = NumbersIteratableDataset()
shard_1.sharding_settings(number_of_shards, 2)
```
you might have something like:
```
number_of_shards = 100
worker_one_shards = list(range(1,33))
worker1 = NumbersIterableDataSet()
worker1.sharding_settings...
worker_two_shards = list(range(33,66))
...
```
I.e you might have 100,000 rows, 100 shards, and 3 "workers." Each worker handles about 33 shards which is ~33,000 rows. Then you just have some extra communication overhead that if one worker is idle you need to move some shards from a busy node to that worker. You can add some rebalancing mechanism so that shards end up being moved to faster workers until they're allocated according to the speed of the hardware.

Going back to the original example, the same thing can probably achieved just by having a larger number of shards per machine.

**Comment by ejguan** [comment]
@zhuzilin We really appreciate your feedback.
Just want to mention a fact that pytorch turns Map-style dataset into iter-style in the DataLoader at the end. All the benefits of MapDataset are the operations over indexes in Sampler. If we have already replaced sampler by directly doing operations over indexes within the pipeline, I don't really see random accessing to dataset is prevented by iter-style datapipe.

For sharding, what's your opinion about the following pipeline.
```py
listfiles_dp = ListFiles(root=data_dir)
shard_dp = Shard(num_workers, worker_index)
loadfiles_dp = LoadFiles(shard_dp)
decoded_images_dp = DecodeImages(loadfiles_dp)
...
```
It definitely requires users more knowledge about the arch to add sharding pipe at the front of pipeline. We probably need to provide the factory pipelines for users to construct them without worry about where they should attach sharding pipe.

And that's the reason I said the memory overhead would be similar fot iter and map in this case. For map-styple, even if you do permutation over indexes, you still have a dict to save index to filepath or whatever.

> This work around would fail if the user have uneven-sized tfrecord files or maybe only one large file, which are common in practice.

It's correct but heavily relying on the meta of TFRecord file to do the random access. It's not common case for users when they are using their own file format like tar, etc.

> In our scenario, the user would save some large dataset on remote file systems like ceph and the latency fluctuation of the FS may affect the training speed a lot. Therefore, we need to cache part of the dataset to local disk to make IO more stable. However, it's hard to implement caching if we only need to store part of the data (The whole dataset is too large (TBs) to be hold in the local disk (GBs)), because we cannot check if a data entry is cached or not. Even if we can, we cannot get the entries that are not cached, since we cannot have random access to the dataset. Therefore, we only implement the caching functionality for pytorch, but not tf at the moment.

I am sorry that I am not sure that I understand what the exact need. Could you elaborate why iter-style datapipe doesn't work in this case?

**Comment by ejguan** [comment]
> @ejguan ListDirFilesIterDataPipe uses generator to yield shard file names, and `__len__` is not implemented only if we pass in the length.
> 
> ```python
>     def __iter__(self) -> Iterator[str] :
>         yield from get_file_pathnames_from_root(self.root, self.masks, self.recursive, self.abspath)
> 
>     def __len__(self):
>         if self.length == -1:
>             raise NotImplementedError
>         return self.length
> ```
> 
> So my question is how sharding works if all the `IterDataPipe` used are of unknown length. Or does it mean we may need pre-compute the number of (shard) files and then inject that to shard settings?

@npuichigo After `ListDirFilesIterDataPipe` yields all file names under a directory, users can choose basically whatever they want to shard these files.
Ohhh, if you are talking about using DataPipe with DistrubutedSampler in DataLoader, which requires `len` to generate random permutation of indexes, you need to wait for some time. We probably will update DataLoader later to make it forward compatible with new DataPipes then you don't need to do sampling for it. All the functionality of sampler will be replaced by other DataPipe without requiring `__len__`.

**Comment by zhuzilin** [comment]
@ejguan 
Sorry that I didn't expain the cache part clearly. Let me try to start with a simpler version. We could accelerate the IO pipeline by cache the data in the memory.

In `tf.data`, the user will:

```python
dataset = tf.data.TFRecordDataset(filenames)
dataset = dataset.cache()
dataset = dataset.repeat()
```

The inner logic of `cache` is similar to:

```python
class CacheDataset(Dataset):
    def __init__(self, input_dataset):
        self.cache = Cache()
        self.input_dataset = input_dataset
        self.index = 0

    def __iter__(self):
        if self.index < len(self.input_dataset):
            data = next(self.input_dataset)
            self.cache[self.index] = data
        else:
            data = self.cache[self.index % len(self.input_dataset)]
        self.index += 1
        return data
```

It will start reuse the cache only after the whole dataset is cached into memory. And that is why we need to put the `repeat` after `cache`, otherwise the `cache` will keep on saving until OOM but never use the cached data.

However, we cannot cache the whole dataset in most cases, so we hope to cache only part of it and for the remains that are not cached, we could still read from the `input_dataset`. Something like this:

```python
class CacheDataset(Dataset):
    def __init__(self, input_dataset, limit):
        # add a size limit to the cache
        self.cache = Cache(limit=limit)
        self.input_dataset = input_dataset
        self.index = 0

    def __iter__(self):
        if self.index in self.cache:
            data = self.cache[self.index % len(self.input_dataset)]
        else:
            # here we need to have random access to the input_dataset
            data = self.input_dataset[self.index]
            if not self.cache.full():
                self.cache[self.index % len(self.input_dataset)] = data
        self.index += 1
        return data
```

As the comment in the above code snippet, when trying to get data directly from the input dataset, we need to rely on random access. (I just found it is still about skipping some records...)

And in the origin comment, our team was trying to add one more hierachy to the caching mechanism. To deal with the remote storage system, we will have:
```
remote storage --|cache part|--> local disk --|cache part|--> memory
```
And this will also need the map-style.

**Comment by zhuzilin** [comment]
@ejguan 
> If we have already replaced sampler by directly doing operations over indexes within the pipeline, I don't really see random accessing to dataset is prevented by iter-style datapipe.

I strongly agree on this and that's why I asked about the `MapDataPipe` in the first place. Because currently, we only have `IterDataPipe` in the master and it does not have a `__getitem__` method for doing operations over indices. I'm curious about how will the index relevant part be merged into the current design. For instance, shall we have a different version for the `MapDataPipe` for operations like `Map`, `Concat`, `Filter`? how will the new `DataLoader`deal with the pipe graph node with index operation and without?

**Comment by VitalyFedyunin** [comment]
The work in `MapDataPipe` is in progress now. We will also to `map` and `concat` but not `filter` for obvious reasons (as indexing becomes unclear or slow).

**Comment by VitalyFedyunin** [comment]
> About a dataset, where sharding results in multiple workers having an uneven amount of data, which can result in "incomplete last batches". Currently, data from different shards is not merged, will this change?
> 
> Related issue: #44108

Yes, it will change as we are going to support mini-batches and would be able to concatenate all 'streams' data.

**Comment by VitalyFedyunin** [comment]
> > Answering sharding question. Let say we have dataset with access to some database of N elements.
> > We need to shard it and yield only 1/X of elements from this database, non overlapping with other
> > shards.
> > ```python
> > class IterateableDataset:
> >     pass
> > 
> > class NumbersIteratableDataset(IterateableDataset):
> >     def is_shardable(self):
> >         return True
> > 
> >     def __init__(self, range = 100):
> >         self.range = range
> > 
> >     def sharding_settings(self, total_num_of_shards, id_of_shard, seed = 0):
> >         # Called before __iter__
> >         self.total_num_of_shards = total_num_of_shards
> >         self.id_of_shard = id_of_shard
> >         self.seed = seed
> > 
> >     def __iter__(self):
> >         for i in range(self.id_of_shard, self.range, self.total_num_of_shards):
> >             yield (i + self.seed) % self.range
> > ```
> > 
> > 
> > Now lets assume X == 3, so we have three consumers (and it doesn't matter how they are separated by physical/logical nodes).
> > For example it can be 1 on shard inside one process and two shards on the different machine.
> > ```python
> > # First
> > number_of_shards = 3
> > shard = NumbersIteratableDataset()
> > shard.sharding_settings(number_of_shards, 0)
> > ```
> > 
> > 
> > ```python
> > # Second
> > number_of_shards = 3
> > shard_0 = NumbersIteratableDataset()
> > shard_0.sharding_settings(number_of_shards, 1)
> > shard_1 = NumbersIteratableDataset()
> > shard_1.sharding_settings(number_of_shards, 2)
> > ```
> > 
> > 
> > So we basically need to share between machines only total number of shards and some cross system incremental id
> > ```python
> > for i in iter(shards_1):
> >     print(i)
> > # 2, 5, ...., 98
> > ```
> > 
> > 
> > It is likely that we will need to have access to other data on next epoch, for this we can introduce seed or epock number to the sharding_settings
> > ```python
> > # Next epoch 
> > for i, shard in enumerate(shards):
> >     shard.sharding_settings(number_of_shards, i, seed = 777 * epoch_number )
> > 
> > for i in iter(shards_2):
> >     print(i)
> > # 79, 82 .... 97, 0, 3, ... 75
> > ```
> 
> @VitalyFedyunin Another question about sharding. Since we may use `ListDirFilesIterDataPipe` to yield shards, sometimes we don't know the actual number of shards. So how to deal with this situation.

Oh, I see my wording is confusing and needs to be corrected in the document. When I wrote 'shards' I meant a number of consumers, not the number of partitions in the source dataset. For sure, having the number of shards divisible by the numbers of consumers is preferred for even load, but not necessary.

**Comment by VitalyFedyunin** [comment]
> Re: Uneven end of data handling
> 
> A common solution is an extra layer of indirection between "shards" and workers (which I guess would be instances of `NumbersIteratableDataset()` in the example). E.g., instead of:

> ...

> Going back to the original example, the same thing can probably achieved just by having a larger number of shards per machine.

Thanks, this is exactly where my documentation is incomplete and require more details to separate shards and workers (and make API cleaner).

**Comment by munael** [comment]
Just curious what you guys think of InfiniBatch. Does it apply here at all? What's different or missing compared to your vision (I admit, couldn't follow the discussion and idea doc all that well 😅).

https://github.com/microsoft/infinibatch

**Comment by VitalyFedyunin** [comment]
> Just curious what you guys think of InfiniBatch. Does it apply here at all? What's different or missing compared to your vision (I admit, couldn't follow the discussion and idea doc all that well 😅).
> 
> https://github.com/microsoft/infinibatch

InfiniBatch looks interesting but doesn't fit our criteria of full backward compatibility and support of map style datasets
